iree/runtime/session.h

```diff
@@ -0,0 +1,224 @@
+
+// Creates a new session forced to use the given |device|.
+// This bypasses any device enumeration performed by the loaded modules but
+// the loaded modules will still verify that the device matches their
+// requirements.
+//
+// A base set of modules may be added by the runtime during creation based on
+// |options| and users may load additional modules - such as the one containing
+// their user code - by using the iree_vm_context_t provided by
+// iree_runtime_session_context.
+//
+// |host_allocator| will be used to allocate the session and any associated
+// resources. |out_session| must be released by the caller.
+IREE_API_EXPORT iree_status_t iree_runtime_session_create_with_device(
+    iree_runtime_instance_t* instance,
+    const iree_runtime_session_options_t* options, iree_hal_device_t* device,
+    iree_allocator_t host_allocator, iree_runtime_session_t** out_session);
+
+// Retains the given |session| for the caller.
+IREE_API_EXPORT void iree_runtime_session_retain(
+    iree_runtime_session_t* session);
+
+// Releases the given |session| from the caller.
+IREE_API_EXPORT void iree_runtime_session_release(
+    iree_runtime_session_t* session);
+
+// Returns the host allocator used to allocate the session and its resources.
+// Callers should use this to allocate resources so that any memory tracking
+// being performed correctly attributes the allocations to the session.
+IREE_API_EXPORT iree_allocator_t
+iree_runtime_session_host_allocator(const iree_runtime_session_t* session);
+
+// Returns the instance the session uses for shared resources.
+IREE_API_EXPORT iree_runtime_instance_t* iree_runtime_session_instance(
+    const iree_runtime_session_t* session);
+
+// Returns the VM context used to load and link modules.
+// The context can be used to perform additional reflection over the loaded
+// modules or load additional modules (if supported).
+IREE_API_EXPORT iree_vm_context_t* iree_runtime_session_context(
+    const iree_runtime_session_t* session);
+
+// Returns the HAL device being used for execution.
+//
+// NOTE: this device will not be available until initialized by a user module
+// and will return NULL if queried prior.
+IREE_API_EXPORT iree_hal_device_t* iree_runtime_session_device(
+    const iree_runtime_session_t* session);
+
+// Returns the device allocator used to allocate compatible buffers.
+// Buffers from other allocators may not be compatible and require importing
+// prior to being usable by the session.
+//
+// NOTE: this device allocator will not be available until initialized by a
+// user module and will return NULL if queried prior.
+IREE_API_EXPORT iree_hal_allocator_t* iree_runtime_session_device_allocator(
+    const iree_runtime_session_t* session);
+
+// Appends the given |module| to the context.
+// The module will be retained by the context.
+//
+// NOTE: only valid if the context is not yet frozen; see
+// iree_runtime_session_create for more information.
```

Comment:
I'm not quite following these NOTEs. I see `iree_runtime_session_create_with_device`, but not `iree_runtime_session_create`, and no mentions of "frozen" contexts?

Reply:
ah yeah, I removed that method until #5724 (as today you have to specify the device) - will fix the note. Freezing contexts is also something not yet implemented so I'll file an issue (today contexts are frozen by default when specifying modules during creation and a manual way to freeze them needs to be added).